### PR TITLE
fix(annotation): rewrite jobs_communication sync to async http communication

### PR DESCRIPTION
**Pull request by Negoje**
Rewrite jobs_communication sync to async http communication

### CODE REVIEW

File: annotation/annotation/microservice_communication/jobs_communication.py [hunk 1]
```diff
@@ -23,49 +22,97 @@
...
+                callback_url, headers=headers, json=json_data
+            ) as response:
+                if response.status != 200:
+                    text = await response.text()
+                    raise JobUpdateException(text)
+        except ClientError as exc:
+            raise JobUpdateException(exc)
+
+
+# def update_job_status_sync(
```
Comment (BogicevicMilan) [inline annotation/annotation/microservice_communication/jobs_communication.py:51]: Please remove this if you don't plan to use it in the future.

File: annotation/annotation/microservice_communication/jobs_communication.py [hunk 2]
```diff
@@ -23,49 +22,97 @@
...
+                url, headers=headers, json=json_data
+            ) as response:
+                if response.status != 200:
+                    text = await response.text()
+                    raise JobUpdateException(text)
+        except ClientError as exc:
+            raise JobUpdateException(exc)
+
+
+# def append_job_categories_sync(
```
Comment (BogicevicMilan) [inline annotation/annotation/microservice_communication/jobs_communication.py:96]: same comment

File: annotation/tests/test_microservices_jobs_communication.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+@pytest.mark.asyncio
+async def test_update_job_status_client_error(job_status_setup):
+    callback_url, status, tenant, token, _ = job_status_setup
+    with patch(
+        "annotation.microservice_communication.jobs_communication."
+        "ClientSession.put",
+        side_effect=ClientError("Connection Failed"),
+    ):
+        with pytest.raises(JobUpdateException):
+            await update_job_status(callback_url, status, tenant, token)
```
[the remark this answers is not in the document]
Comment (Negoje) [reply annotation/tests/test_microservices_jobs_communication.py:129]: You should be able to install pytest-asyncio now and all test should work

**Review by UkiStefanovic — APPROVED** [review]
Looks good

File: annotation/annotation/microservice_communication/jobs_communication.py [hunk 4]
```diff
@@ -23,49 +22,52 @@
...
+    json_data = {"status": status}
+    timeout = ClientTimeout(total=5)
+    async with ClientSession(timeout=timeout) as session:
+        try:
+            async with session.put(
+                callback_url, headers=headers, json=json_data
+            ) as response:
+                if response.status != 200:
+                    text = await response.text()
+                    raise JobUpdateException(text)
```
Comment (cakeinsauce) [inline annotation/annotation/microservice_communication/jobs_communication.py:45]:
```suggestion
                    raise JobUpdateException(await response.text())
```

File: annotation/annotation/microservice_communication/jobs_communication.py [hunk 5]
```diff
@@ -23,49 +22,52 @@
...
+    json_data = {"categories_append": categories}
+    timeout = ClientTimeout(total=5)
+    async with ClientSession(timeout=timeout) as session:
+        try:
+            async with session.put(
+                url, headers=headers, json=json_data
+            ) as response:
+                if response.status != 200:
+                    text = await response.text()
+                    raise JobUpdateException(text)
```
Comment (cakeinsauce) [inline annotation/annotation/microservice_communication/jobs_communication.py:68]:
```suggestion
                    raise JobUpdateException(await response.text())
```

File: annotation/pyproject.toml [hunk 6]
```diff
@@ -20,13 +20,16 @@ kafka-python = "^2.0.2"
 sqlalchemy-utils = "^0.38.3"
 lru-dict = "1.2.0"
 badgerdoc-storage = { path = "../lib/badgerdoc_storage" }
+aiohttp = "^3.10.5"
+pytest = "^8.2"
```
Comment (cakeinsauce) [inline annotation/pyproject.toml:24]: pytest is not a project dependency, it is dev only

Comment (cakeinsauce) [reply annotation/pyproject.toml:24]: remove it and update lock file, please

File: annotation/tests/test_distribution.py [hunk 7]
```diff
@@ -1856,7 +1856,8 @@ def test_distribute_extensive_coverage(
         ([], JobStatusEnumSchema.finished),
     ),
 )
-def test_redistribute(
+@pytest.mark.asyncio
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1859]: Remove this decorator from all tests and add configuration to pyproject.toml for pytest to automatically include it
https://pytest-asyncio.readthedocs.io/en/latest/reference/configuration.html

File: annotation/tests/test_microservices_jobs_communication.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    token = "example_token"
+    mock_response = AsyncMock()
+    mock_response.__aenter__.return_value = mock_response
+    mock_response.__aexit__.return_value = None
+
+    yield (job_id, categories, tenant, token, mock_response)
+
+
+@pytest.mark.asyncio
+async def test_append_job_categories_success(setup_append_job_categories):
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:27]: typing, please

File: annotation/tests/test_microservices_jobs_communication.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+@pytest.fixture
+def setup_append_job_categories():
+    job_id = "123"
+    categories = ["Engineering", "Management"]
+    tenant = "example_tenant"
+    token = "example_token"
+    mock_response = AsyncMock()
+    mock_response.__aenter__.return_value = mock_response
+    mock_response.__aexit__.return_value = None
+
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:22]:
```suggestion
```

File: annotation/tests/test_microservices_jobs_communication.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        "annotation.microservice_communication.jobs_communication."
+        "ClientSession.put",
+        return_value=mock_response,
+    ):
+        await append_job_categories(job_id, categories, tenant, token)
+
+
+@pytest.mark.asyncio
+async def test_append_job_categories_failure_non_200(
+    setup_append_job_categories,
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:47]: typing, please

File: annotation/tests/test_microservices_jobs_communication.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        "annotation.microservice_communication.jobs_communication."
+        "ClientSession.put",
+        return_value=mock_response,
+    ):
+        with pytest.raises(JobUpdateException):
+            await append_job_categories(job_id, categories, tenant, token)
+
+
+@pytest.mark.asyncio
+async def test_append_job_categories_client_error(setup_append_job_categories):
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:70]: typing, please

File: annotation/tests/test_microservices_jobs_communication.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    token = "example_token"
+    mock_response = AsyncMock()
+    mock_response.__aenter__.return_value = mock_response
+    mock_response.__aexit__.return_value = None
+
+    yield (callback_url, status, tenant, token, mock_response)
+
+
+@pytest.mark.asyncio
+async def test_update_job_status_success(job_status_setup):
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:95]: typing, please

File: annotation/tests/test_microservices_jobs_communication.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    with patch(
+        "annotation.microservice_communication.jobs_communication."
+        "ClientSession.put",
+        return_value=mock_response,
+    ):
+        await update_job_status(callback_url, status, tenant, token)
+
+
+@pytest.mark.asyncio
+async def test_update_job_status_failure_non_200(job_status_setup):
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:107]: typing, please

File: annotation/tests/test_microservices_jobs_communication.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        "annotation.microservice_communication.jobs_communication."
+        "ClientSession.put",
+        return_value=mock_response,
+    ):
+        with pytest.raises(JobUpdateException):
+            await update_job_status(callback_url, status, tenant, token)
+
+
+@pytest.mark.asyncio
+async def test_update_job_status_client_error(job_status_setup):
```
Comment (cakeinsauce) [inline annotation/tests/test_microservices_jobs_communication.py:121]: typing, please

**Review by cakeinsauce — CHANGES_REQUESTED** [review]

File: annotation/pytest.ini [hunk 15]
```diff
@@ -2,3 +2,4 @@
 markers =
     integration: test to demonstrate, that different parts of systems work together, requires dependencies on outside systems (like db)
     unittest: small piece of code, does not depend on outside systems (like db)
+asyncio_mode = auto
```
Comment (cakeinsauce) [inline annotation/pytest.ini:5]: add new line, please

**Review by cakeinsauce — APPROVED** [review]
Great, test it in fully working env, please

**Pull request merged**